src/H5Dcompact.c

```diff
@@ -239,6 +259,48 @@ H5D__compact_io_init(const H5D_io_info_t *io_info, const H5D_type_info_t H5_ATTR
     FUNC_LEAVE_NOAPI(SUCCEED)
 } /* end H5D__compact_io_init() */
 
+/*-------------------------------------------------------------------------
+ * Function:    H5D__compact_readvv_special_mem_cb
+ *
+ * Purpose:     Callback operator for H5D__compact_readvv() to send a
+ *              memory copy request to the underlying file driver.
+ *
+ * Return:      Non-negative on success/Negative on failure
+ *
+ *-------------------------------------------------------------------------
+ */
+static herr_t
+H5D__compact_readvv_special_mem_cb(hsize_t dst_off, hsize_t src_off, size_t len, void *_udata)
+{
+    H5D_compact_readvv_special_mem_ud_t *udata = (H5D_compact_readvv_special_mem_ud_t *)_udata;
+    H5FD_ctl_memcpy_args_t               op_args;
+    uint64_t                             op_flags;
+    H5FD_t *                             file_handle = NULL;
+    herr_t                               ret_value   = SUCCEED;
+
+    FUNC_ENTER_STATIC
+
+    /* Retrieve lower-level file handle for I/O */
+    if (H5F_shared_get_file_handle(udata->f_sh, &file_handle) < 0)
```

Comment:
Here, we need a handle to the shared file structure's H5FD_t pointer. However, the H5F package hides this behind an incomplete typedef, so we can't just directly access it through udata->f_sh->lf. We have to create a routine in H5Fquery.c to give this back to us.